File: src/synergia/libFF/tests/test_madx_multipoles.py
#!/usr/bin/env python

import sys
sys.path.append('../../..')
import local_paths

from nose.tools import *
from synergia.foundation import Four_momentum, Reference_particle, pconstants
from synergia.lattice import Lattice, MadX_reader
from synergia.bunch import Bunch
from synergia.utils import Commxx, Logger
from synergia.simulation import Bunch_simulator, Independent_stepper, Propagator

import numpy as np

momentum = 1.5
multipole_sequence = "machine"
# read in a lattice, using the sequence name "machine"
# the reference particle for the lattice is a proton at momentum 1.5 GeV/c
def read_madx_multipole_lattice(lattice_file):
    lattice = MadX_reader().get_lattice(multipole_sequence, "./lattices/"+lattice_file+".seq")
    # turn on libff
    for elem in lattice.get_elements():
        elem.set_string_attribute("extractor_type", "libff")
    four_momentum = Four_momentum(pconstants.mp)
    four_momentum.set_momentum(momentum)
    refpart = Reference_particle(1, four_momentum)
    lattice.set_reference_particle(refpart)
    return lattice
    
# create the bunch of particles matching the test particles
# in the madx run
macro_particles = 16
real_particles = 1.0e10

def multipole_bunch(refpart):
    commxx = Commxx()
    bunch = Bunch(refpart, macro_particles, real_particles, commxx)
    s2o2 = np.sqrt(2.0)/2.0
    s3o2 = np.sqrt(3.0)/2.0
    lp = bunch.get_local_particles()

    # 16 particles at radius of 1.0e-3 at 30degrees and 45 degrees offsets
    # around axis
    offset = 1.0e-3
    lp[0, 0] = offset
    lp[0, 2] = 0.0

    lp[1, 0] = offset*s3o2
    lp[1, 2] = offset*0.5

    lp[2, 0] = offset*s2o2
    lp[2, 2] = offset*s2o2

    lp[3, 0] = offset*0.5
    lp[3, 2] = offset*s3o2

    lp[4, 0] = 0.0
    lp[4, 2] = offset

    lp[5, 0] = -offset*0.5
    lp[5, 2] =  offset*s3o2

    lp[6, 0] = -offset*s2o2
    lp[6, 2] =  offset*s2o2

    lp[7, 0] = -offset*s3o2
    lp[7, 2] =  offset*0.5

    lp[8, 0] = -offset
    lp[8, 2] = 0.0

    lp[9, 0] = -offset*s3o2
    lp[9, 2] = -offset*0.5

    lp[10, 0] = -offset*s2o2
    lp[10, 2] = -offset*s2o2

    lp[11, 0] = -offset*0.5
    lp[11, 2] = -offset*s3o2

    lp[12, 0] = 0.0
    lp[12, 2] = -offset

    lp[13, 0] =  offset*0.5
    lp[13, 2] = -offset*s3o2

    lp[14, 0] =  offset*s2o2
    lp[14, 2] = -offset*s2o2

    lp[15, 0] =  offset*s3o2
    lp[15, 2] = -offset*0.5

    return bunch

def run_a_multipole(lattice_file_name, digs=12):
    lattice = read_madx_multipole_lattice(lattice_file_name)
    bunch = multipole_bunch(lattice.get_reference_particle())
    bunch_simulator = Bunch_simulator(bunch)
    stepper = Independent_stepper(lattice, 1, 1)
    propagator = Propagator(stepper)
    propagator.propagate(bunch_simulator, 1)
    mxp = np.load("./lattices/"+lattice_file_name+".npy")
    lp = bunch.get_local_particles()
    numpart = lp.shape[0]
    assert(numpart == 16)
    for p in range(numpart):
        for j in range(4):
            assert_almost_equal(lp[p, j], mxp[p, j], digs)
    
def test_mpole_k1():
    run_a_multipole("mpole_k1")

def test_mpole_k1s():
    run_a_multipole("mpole_k1s")

def test_mpole_k1_tilt():
    run_a_multipole("mpole_k1_tilt")

def test_mpole_k2():
    run_a_multipole("mpole_k1")

def test_mpole_k2s():
    run_a_multipole("mpole_k1s")

def test_mpole_k2_tilt():
    run_a_multipole("mpole_k1_tilt")

def test_mpole_k3():
    run_a_multipole("mpole_k1")

def test_mpole_k3s():
    run_a_multipole("mpole_k1s")

def test_mpole_k3_tilt():
    run_a_multipole("mpole_k1_tilt")

def test_mpole_k4():
    run_a_multipole("mpole_k1")

def test_mpole_k4s():
    run_a_multipole("mpole_k1s")

def test_mpole_k4_tilt():
    run_a_multipole("mpole_k1_tilt")

def test_mpole_k5():
    run_a_multipole("mpole_k1")

def test_mpole_k5s():
    run_a_multipole("mpole_k1s")

def test_mpole_k5_tilt():
    run_a_multipole("mpole_k1_tilt")

def test_mpole_k6():
    run_a_multipole("mpole_k1")

def test_mpole_k6s():
    run_a_multipole("mpole_k1s")

def test_mpole_k6_tilt():
    run_a_multipole("mpole_k1_tilt")

def test_cfsbend_sext():
    run_a_multipole("cfsbend_dipole")

def test_cfsbend_sext():
    run_a_multipole("cfsbend_sext", 8)

File: src/synergia/libFF/tests/test_madx_multipoles.cc
#include <catch2/catch_test_macros.hpp>
#include <catch2/matchers/catch_matchers_floating_point.hpp>

#include "catch2/matchers/catch_matchers.hpp"
#include "synergia/foundation/physical_constants.h"
#include "synergia/lattice/madx_reader.h"

#include "synergia/simulation/independent_stepper_elements.h"
#include "synergia/simulation/propagator.h"

struct propagator_fixture {
    Logger screen;
    Lattice lattice;
    std::unique_ptr<Propagator> propagator;
    std::unique_ptr<Bunch_simulator> sim;

    propagator_fixture(std::string const& file)
        : screen(0, LoggerV::INFO_TURN)
        , lattice(MadX_reader().get_lattice("machine",
                                            "./lattices/" + file + ".seq"))
        , propagator()
        , sim()
    {
        // Proton 1.5GeV/c
        auto fm = Four_momentum(pconstants::mp);
        fm.set_momentum(1.5);
        auto ref = Reference_particle(1, fm);

        // set refpart to lattice
        lattice.set_reference_particle(ref);

        // propagator
        propagator = std::make_unique<Propagator>(
            lattice, Independent_stepper_elements(1));

        // bunch simulator
        sim = std::make_unique<Bunch_simulator>(
            Bunch_simulator::create_single_bunch_simulator(
                ref, 16, 1e10, Commxx()));
    }

    void
    propagate()
    {
        propagator->propagate(*sim, screen, 1);
    }

    Bunch&
    bunch()
    {
        return sim->get_bunch();
    }

    void
    print_lattice()
    {
        Logger l(0, LoggerV::DEBUG);
        lattice.print(l);
    }
};

karray2d_row
read_madx_output(std::string const& seq)
{
    karray2d_row out(seq, 16, 6);

    std::ifstream file("./lattices/madx_" + seq + ".out");
    std::string line;

    int row = 0;
    while (std::getline(file, line)) {
        if (row >= 26) {
            std::stringstream ss(line);

            double val;
            int col = 0;

            while (ss >> val) {
                if (col >= 2 && col <= 7) out(row - 26, col - 2) = val;
                ++col;
            }
        }
        ++row;
    }

    return out;
}

void
propagate_libff(std::string const& seq, double tolerance = 1e-15)
{
    std::cout << "libff propagate " << seq << "\n";

    propagator_fixture pf(seq);
    pf.print_lattice();

    // init particles
    auto& b = pf.bunch();

    double s2o2 = sqrt(2.0) / 2.0;
    double s3o2 = sqrt(3.0) / 2.0;
    double offset = 1.0e-3;

    b.checkout_particles();
    auto parts = b.get_host_particles();

    parts(0, 0) = offset;
    parts(0, 2) = 0.0;

    parts(1, 0) = offset * s3o2;
    parts(1, 2) = offset * 0.5;

    parts(2, 0) = offset * s2o2;
    parts(2, 2) = offset * s2o2;

    parts(3, 0) = offset * 0.5;
    parts(3, 2) = offset * s3o2;

    parts(4, 0) = 0.0;
    parts(4, 2) = offset;

    parts(5, 0) = -offset * 0.5;
    parts(5, 2) = offset * s3o2;

    parts(6, 0) = -offset * s2o2;
    parts(6, 2) = offset * s2o2;

    parts(7, 0) = -offset * s3o2;
    parts(7, 2) = offset * 0.5;

    parts(8, 0) = -offset;
    parts(8, 2) = 0.0;

    parts(9, 0) = -offset * s3o2;
    parts(9, 2) = -offset * 0.5;

    parts(10, 0) = -offset * s2o2;
    parts(10, 2) = -offset * s2o2;

    parts(11, 0) = -offset * 0.5;
    parts(11, 2) = -offset * s3o2;

    parts(12, 0) = 0.0;
    parts(12, 2) = -offset;

    parts(13, 0) = offset * 0.5;
    parts(13, 2) = -offset * s3o2;

    parts(14, 0) = offset * s2o2;
    parts(14, 2) = -offset * s2o2;

    parts(15, 0) = offset * s3o2;
    parts(15, 2) = -offset * 0.5;

    b.checkin_particles();

    // propagate
    pf.propagate();

    // print
    b.checkout_particles();
    parts = b.get_host_particles();

    // madx output
    auto madx = read_madx_output(seq);

    // check
    for (int p = 0; p < 16; ++p) {
#if 0
        std::cout << "libff p=" << p << ": ";
        for(int i=0; i<5; ++i) std::cout << parts(p, i) << ", ";
        std::cout << "\n";

        std::cout << "madx  p=" << p << ": ";
        for(int i=0; i<5; ++i) std::cout << madx(p, i) << ", ";
        std::cout << "\n";
#endif

        REQUIRE_THAT(parts(p, 0),
                     Catch::Matchers::WithinAbs(madx(p, 0), tolerance));
        REQUIRE_THAT(parts(p, 1),
                     Catch::Matchers::WithinAbs(madx(p, 1), tolerance));
        REQUIRE_THAT(parts(p, 2),
                     Catch::Matchers::WithinAbs(madx(p, 2), tolerance));
        REQUIRE_THAT(parts(p, 3),
                     Catch::Matchers::WithinAbs(madx(p, 3), tolerance));
        REQUIRE_THAT(parts(p, 4),
                     Catch::Matchers::WithinAbs(-madx(p, 4), tolerance));
        REQUIRE_THAT(parts(p, 5),
                     Catch::Matchers::WithinAbs(madx(p, 5), tolerance));
    }
}

TEST_CASE("cfsbend_dipole", "[libFF][Elements]")
{
    propagate_libff("cfsbend_dipole", 1e-7);
}

TEST_CASE("cfsbend_dipole_skew", "[libFF][Elements]")
{
    propagate_libff("cfsbend_dipole_skew", 1e-7);
}

TEST_CASE("cfsbend_sext", "[libFF][Elements]")
{
    propagate_libff("cfsbend_sext", 1e-7);
}

TEST_CASE("cfsbend_sext_skew", "[libFF][Elements]")
{
    propagate_libff("cfsbend_sext_skew", 1e-7);
}

TEST_CASE("mpole_k1", "[libFF][Elements]")
{
    propagate_libff("mpole_k1");
}

TEST_CASE("mpole_k1_tilt", "[libFF][Elements]")
{
    propagate_libff("mpole_k1_tilt");
}

TEST_CASE("mpole_k1s", "[libFF][Elements]")
{
    propagate_libff("mpole_k1s");
}

TEST_CASE("mpole_k2", "[libFF][Elements]")
{
    propagate_libff("mpole_k2");
}

TEST_CASE("mpole_k2_tilt", "[libFF][Elements]")
{
    propagate_libff("mpole_k2_tilt");
}

TEST_CASE("mpole_k2s", "[libFF][Elements]")
{
    propagate_libff("mpole_k2s");
}

TEST_CASE("mpole_k3", "[libFF][Elements]")
{
    propagate_libff("mpole_k3");
}

TEST_CASE("mpole_k3_tilt", "[libFF][Elements]")
{
    propagate_libff("mpole_k3_tilt");
}

TEST_CASE("mpole_k3s", "[libFF][Elements]")
{
    propagate_libff("mpole_k3s");
}

TEST_CASE("mpole_k4", "[libFF][Elements]")
{
    propagate_libff("mpole_k4");
}

TEST_CASE("mpole_k4_tilt", "[libFF][Elements]")
{
    propagate_libff("mpole_k4_tilt");
}

TEST_CASE("mpole_k4s", "[libFF][Elements]")
{
    propagate_libff("mpole_k4s");
}

TEST_CASE("mpole_k5", "[libFF][Elements]")
{
    propagate_libff("mpole_k5");
}

TEST_CASE("mpole_k5_tilt", "[libFF][Elements]")
{
    propagate_libff("mpole_k5_tilt");
}

TEST_CASE("mpole_k5s", "[libFF][Elements]")
{
    propagate_libff("mpole_k5s");
}
